tests/components/knx/test_init.py

```diff
@@ -102,6 +184,39 @@ async def test_init_connection_handling(
 
     assert hass.data.get(KNX_DOMAIN) is not None
 
-    assert (
-        hass.data[KNX_DOMAIN].connection_config().__dict__ == connection_config.__dict__
+    original_connection_config = (
+        hass.data[KNX_DOMAIN].connection_config().__dict__.copy()
     )
+    del original_connection_config["secure_config"]
+
+    connection_config_dict = connection_config.__dict__.copy()
+    del connection_config_dict["secure_config"]
+
+    assert original_connection_config == connection_config_dict
+
+    if connection_config.secure_config is not None:
+        assert (
+            hass.data[KNX_DOMAIN].connection_config().secure_config.knxkeys_password
```

Comment:
Please don't access hass.data in tests.

Reply:
How can we access this otherwise?

Reply:
We don't really care about internal details. We should only assert library interactions or core state or some persistent state eg in storage. What is this case?

Reply:
This tests that the connection config used in the library is properly built based on the data in the config entry.

See https://github.com/home-assistant/core/blob/dev/homeassistant/components/knx/__init__.py#L381-L437

Reply:
Then we should patch the library and assert the call arguments of the mock.